internal/stream/stream_format.go

```diff
@@ -94,6 +107,33 @@ func (sf *streamFormat) writeUnitInner(s *Stream, medi *description.Media, u uni
 
 	atomic.AddUint64(s.bytesReceived, size)
 
+	if sf.gopCache && medi.Type == description.MediaTypeVideo {
+		if isKeyFrame(u) {
+			if s.CachedUnits == nil {
+				// Empty the cache and enable caching
+				s.CachedUnits = []unit.Unit{}
+			} else {
+				// Keep the last packets that were used to generate the key frame. This is to send a full key frame in the RTSP stream.
+				lastPTS := s.CachedUnits[len(s.CachedUnits)-1].GetPTS()
+				i := len(s.CachedUnits)
+				for ; i > 0; i-- {
+					if s.CachedUnits[i-1].GetPTS() != lastPTS {
+						break
+					}
+				}
+				s.CachedUnits = s.CachedUnits[i:]
+			}
+
+		}
+		if s.CachedUnits != nil {
```

Comment:
You probably don't need these checks, a `nil` typed slice is valid to work with:

```go
var a []string
len(a) // 0 
a = append(a, "X")
len(a) // 1
a = nil 
len(a) // 0
a = append(a, "Z")
len(a) // 1

Reply:
Maps on the other hand do need to be initialized before you access them (`len` still works on `nil` typed maps)

Reply:
This is a "trick" I'm using, I let `s.CachedUnits` be `nil` until the stream receives a key frame, this way, we do not waste memory storing p-frames.